File: edudoc/src/app/model/dtos/encounter-data.ts
import { ICptCodeWithMinutesDto } from '@model/interfaces/custom/encounter-response.dto';
import { IGoal } from '@model/interfaces/goal';
import { IMethod } from '@model/interfaces/method';

export type IBasicEncounterLineData = {
    Kind: 'basic';
    EncounterNumber: string;
    EncounterDate: string;
    ServiceType: string;
    Status: string;
    Grouping: number;
    AdditionalStudents: number;
    StartTime: string;
    EndTime: string;
    TotalMinutes: number;
    IsTelehealth: boolean;
};

export type IDetailedEncounterLineData = Omit<IBasicEncounterLineData, 'Kind'> & {
    Kind: 'detailed';
    ReasonForService: string;
    ReasonForDeviation: string;
    Methods: IMethod[];
    Goals: IGoal[];
    ProcedureCodes: ICptCodeWithMinutesDto[];
    Notes: string;
    EntryDate: string;
};

export type IEncounterLineData = IBasicEncounterLineData | IDetailedEncounterLineData;

export type EncounterType = IEncounterLineData['Kind'];

export type IEncounterDistrictData<T extends IEncounterLineData> = {
    DistrictName: string;
    GroupData: IEncounterGroupData<T>[];
};
export type IBasicEncounterDistrictData = IEncounterDistrictData<IBasicEncounterLineData>;
export type IDetailedEncounterDistrictData = IEncounterDistrictData<IDetailedEncounterLineData>;

export type IEncounterGroupData<T extends IEncounterLineData> = {
    ProviderName: string;
    StudentInfo: string;
    IEPStartDate: string;
    TotalMinutes: number;
    LineData: T[];
};

export type IEncounterFullData<T extends IEncounterLineData> = {
	GroupData: Omit<IEncounterGroupData<T>, 'LineData'>,
    DistrictData: Omit<IEncounterDistrictData<T>, 'GroupData'>,
    LineData: T;
}

export type IBasicEncounterFullData = IEncounterFullData<IBasicEncounterLineData>;
export type IDetailedEncounterFullData = IEncounterFullData<IDetailedEncounterLineData>;

File: edudoc/src/Model/Custom/EncounterData.cs
using System.Data.SqlClient;
using System.Data.SqlClient;

using System;
using System.Collections.Generic;
using Model.DTOs;

namespace Model.Custom
{

    public class EncounterDistrictData<TLineData>
    {
        public string DistrictName { get; set; }
        public IList<EncounterGroupData<TLineData>> GroupData { get; set; }
    }

    public class EncounterGroupData<TLineData>
    {
        public string ProviderName { get; set; }
        public string StudentInfo { get; set; }
        public string IEPStartDate { get; set; }
        public int TotalMinutes { get; set; }
        public IList<TLineData> LineData { get; set; }
    }

    public class BasicEncounterLineData
    {
        public string Kind {get;set;} = "basic";
        public string EncounterNumber { get; set; }
        public string Status { get; set; }
        public int Grouping { get; set; }
        public int AdditionalStudents { get; set; }
        public string EncounterDate { get; set; }
        public string ServiceType { get; set; }
        public string StartTime { get; set; }
        public string EndTime { get; set; }
        public int TotalMinutes { get; set; }
        public bool IsTelehealth { get; set; }

    }

    public class DetailedEncounterLineData : BasicEncounterLineData
    {
        public new string Kind {get;set;} = "detailed";
        public string ReasonForService { get; set; }
        public string ReasonForDeviation { get; set; }
        public List<Method> Methods { get; set; }
        public List<Goal> Goals { get; set; }
        public List<CptCodeWithMinutesDto> ProcedureCodes { get; set; }
        public string Notes { get; set; }
        public string EntryDate { get; set; }

    }
}
